### ISSUE TITLE
About structrue-enhanced self-attention

### ISSUE THREAD

**Issue by mqEpiphany** [zjunlp/Relphormer#7]
I saw in your paper that the adjacency matrix is generated from the central triplet as structural information and added to self-attention, because your code is very confusing and I can't find the corresponding code, can you tell me the code for this part where is it？
![image](https://user-images.githubusercontent.com/79396680/206974770-7660ed6b-699e-45d2-af54-31e7f6f55ff5.png)

**Comment by bizhen46766** [comment]
Hi, in this code version, we directly put strucural information as adjacency matrix in the self-attention module for simplicity.
It is easy to put structural encoder (an extra linear to encode the power of attention matrix).
Because the structural encoder is actually like a hyper-parameter and you can choose to use this module which is dependable on the actual knowledge graph you choose.
We will complete this part in the future version and also release the code version for relation prediction.

**Comment by bizhen46766** [comment]
you can find the implementation in the BertSelfAttention module (321 line.) in huggingface_relformer.py,

**Comment by mqEpiphany** [comment]
Hi, thank you for your patient reply. I also want to ask why I can't enter BertSelfAttention in huggingface_relformer.py when I start from the main function for single-step debugging?

**Comment by bizhen46766** [comment]
Did you first pre-train the model for initialization and then load the pre-trained model?
The hyper-parameters of two stage are different and the second stage will import the huggingface_relformer.py.

**Comment by mqEpiphany** [comment]
When performing pre-training or entity prediction, some files need to be downloaded, but it is very slow, so I downloaded it from https://huggingface.co/bert-base-uncased, am I right? The following is the hyperparameter setting for my running entity prediction.
![image](https://user-images.githubusercontent.com/79396680/206999932-d0d02e96-ba48-412e-ad5e-9016c5e73527.png)

**Comment by bizhen46766** [comment]
Yes, that is right. I think I find the different point. 
You should run the "main.py" in the "Relphormer" directory instead of the same file in the "pre-train" directory.

**Comment by mqEpiphany** [comment]
Sorry, I didn't present clearly, I just put the pre-trained model bert-base-uncased in the "pretrain" directory, this is run the "main.py" in the "Relphormer" directory.
![image](https://user-images.githubusercontent.com/79396680/207211669-aff0760d-ba37-41b8-bcc3-756e671fab45.png)

**Comment by bizhen46766** [comment]
ok, can you check the program by putting the debug point in [[from models.huggingface_relformer import [BertForMaskedLM]]](https://github.com/zjunlp/Relphormer/blob/7ea60b54c8b7dbcf9a67241e0e55983eacbe5260/models/model.py#L2)
It imports the "models.huggingface_relformer" file.

**Comment by mqEpiphany** [comment]
Sorry, I can only find the class BertForMaskedLM, but I can't find the code [from models.huggingface_relformer import [BertForMaskedLM]]

**Comment by bizhen46766** [comment]
Can you directly import the huggingface_relformer.py file? Because we rewrite the BertForMaskedLM class in this file.

**Comment by mqEpiphany** [comment]
Do you mean to create a new file and try to import the huggingface_relformer.py file?

**Comment by bizhen46766** [comment]
Yes, you can try do that in the second training stage and it will also make sense.

**Comment by mqEpiphany** [comment]
Single-step debugging of the main function to trainer.fit(lit_model, datamodule=data) will start training, unable to enter huggingface_relformer.py

**Comment by bizhen46766** [comment]
You can find the line which imports the model class.  
https://github.com/zjunlp/Relphormer/blob/7ea60b54c8b7dbcf9a67241e0e55983eacbe5260/main.py#L51
`model_class = _import_class(f"models.{temp_args.model_class}")`

**Comment by mqEpiphany** [comment]
The hyperparameter setting of args.model_class is BertKGC. Will the model in huggingface_relformer.py be called during the execution of BertKGC?

**Comment by bizhen46766** [comment]
https://github.com/zjunlp/Relphormer/blob/7ea60b54c8b7dbcf9a67241e0e55983eacbe5260/models/model.py#L1-L7
The model_class BertKGC is inherited from BertForMaskedLM in huggingface_relformer.py

**Comment by mqEpiphany** [comment]
Doesn't the attention_mask in the BertSelfAttention module (line 321) in huggingface_relformer.py actually build? I saw that attention_mask first came from the image below:
![image](https://user-images.githubusercontent.com/79396680/208230518-a6570bd9-ccda-41f0-a7fe-b3e193369b90.png)

**Comment by bizhen46766** [comment]
Hi, we generate the attention mask for each center triple and you can find in the input of the module.

**Comment by mqEpiphany** [comment]
Can you tell me exactly where in the input module?

**Comment by bizhen46766** [comment]
https://github.com/zjunlp/Relphormer/blob/7ea60b54c8b7dbcf9a67241e0e55983eacbe5260/data/processor.py#L319-L327
